File: Rcurses/R/border.r
#' draw a box around the edges of a window
#'
#' @inheritParams window_params
#' @param ls	left side
#' @param rs	right side
#' @param ts 	top side
#' @param bs	bottom side
#' @param tl	top left-hand corner
#' @param tr	top right-hand corner
#' @param bl	bottom left-hand corner
#' @param br	bottom right-hand corner
#' @export
#' @family border
border <- function(win = stdscr(),
  ls = 0, rs = 0, ts = 0, bs = 0,
  tl = 0, tr = 0, bl = 0, br = 0) {

  .code <- function(x) {
    if (is.character(x)) x <- charToRaw(x)[1]
    as.integer(x)
  }

  .call('_wborder', win,
    .code(ls),
    .code(rs),
    .code(ts),
    .code(bs),
    .code(tl),
    .code(tr),
    .code(bl),
    .code(br))
  invisible()
}


#' draw a box around a window
#'
#' @inheritParams window_params
#' @export
#' @family border
box <- function(win = stdscr()) {
  .call('_box00', win)
  invisible()
}


#' draw  a  vertical  (top  to bottom) line
#'
#'  The vline and wvline functions draw  a  vertical  (top  to
#'        bottom) line using ch starting at the current cursor posi-
#'        tion in the window.  The current cursor  position  is  not
#'        changed.   The  line  is  at most n characters long, or as
#'        many as fit into the window.
#'
#' @inheritParams window_params
#' @param ch 		a character to draw the line with
#' @param n			the max nb of characters to draw
#' @export
#' @family border
wvline <- function(win, ch, n) {
  if (length(ch) != 1 || nchar(ch) != 1)
    stop('give a single character')
  status <- as.logical(.call('_wvline', win, as.character(ch), as.integer(n)))

  invisible(status)
}

#' draw  a  vertical  (top  to bottom) line
#'
#'  The vline and wvline functions draw  a  vertical  (top  to
#'        bottom) line using ch starting at the current cursor posi-
#'        tion in the window.  The current cursor  position  is  not
#'        changed.   The  line  is  at most n characters long, or as
#'        many as fit into the window.
#'
#' @inheritParams window_params
#' @param ch 		a character to draw the line with
#' @param n			the max nb of characters to draw
#' @export
#' @family border
wvline <- function(win, ch, n) {
  if (length(ch) != 1 || nchar(ch) != 1)
    stop('give a single character')
  status <- as.logical(.call('_wvline', win, as.character(ch), as.integer(n)))

  invisible(status)
}



#' vline
#'
#' @inheritParams wvline
#' @export
#' @family border
vline <- function(ch, n, win = stdscr()) {
  wvline(win, ch, n)
}

File: Rcurses/src/border.c
#include "rcurses.h"

SEXP _wborder(SEXP _win,
              SEXP ls, SEXP rs, SEXP ts, SEXP bs,
              SEXP tl, SEXP tr, SEXP bl, SEXP br)
{
  if (!R_ExternalPtrAddr(_win)) return R_NilValue;

  wborder(
      (WINDOW*)R_ExternalPtrAddr(_win),
      INTEGER(ls)[0],
      INTEGER(rs)[0],
      INTEGER(ts)[0],
      INTEGER(bs)[0],
      INTEGER(tl)[0],
      INTEGER(tr)[0],
      INTEGER(bl)[0],
      INTEGER(br)[0]
  );

  return R_NilValue;
}


SEXP _box00(SEXP _win) {
  if (!R_ExternalPtrAddr(_win)) return R_NilValue;
  box((WINDOW*)R_ExternalPtrAddr(_win), 0, 0);
  return R_NilValue;
}

SEXP _wvline(SEXP win, SEXP ch, SEXP n) {
  if (!R_ExternalPtrAddr(win)) return R_NilValue;
  int status = wvline((WINDOW*)R_ExternalPtrAddr(win),
                      CHAR(STRING_ELT(ch, 0))[0], INTEGER(n)[0]);
  return create_status(status);
}

SEXP _whline(SEXP win, SEXP ch, SEXP n) {
  if (!R_ExternalPtrAddr(win)) return R_NilValue;
  int status = whline((WINDOW*)R_ExternalPtrAddr(win),
                      CHAR(STRING_ELT(ch, 0))[0], INTEGER(n)[0]);
  return create_status(status);
}
